overlord/snapstate/handlers.go

```diff
@@ -892,113 +886,195 @@ func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
+	if _, ok := refreshHints[snapName]; !ok {
+		return fmt.Errorf("cannot reset the monitoring field for %q in \"refresh-candidates\"", snapName)
+	}
 
+	refreshHints[snapName].Monitoring = false
+	st.Set("refresh-candidates", refreshHints)
 	return nil
 }
 
-func continueRefreshOnSnapClose(st *state.State, instanceName string, done <-chan string, abort <-chan bool) {
-	continueAutoRefresh := false
+func continueRefreshOnSnapClose(st *state.State, snapsup *SnapSetup, done <-chan string, abort <-chan bool) {
+	snapName := snapsup.InstanceName()
+
+	var aborted bool
 	select {
 	case <-done:
-		continueAutoRefresh = true
 	case <-abort:
+		aborted = true
 	}
 
 	st.Lock()
 	defer st.Unlock()
 
-	monitoredSnaps := snapMonitoring(st)
-	if monitoredSnaps == nil {
-		// shouldn't happen except for programmer error
-		logger.Noticef("cannot find monitoring state for snap %q", instanceName)
-	} else {
-		delete(monitoredSnaps, instanceName)
-
-		if len(monitoredSnaps) == 0 {
-			// use nil to delete entry but must be nil type (can't be map var set to nil)
-			updateMonitoringState(st, nil)
-		} else {
-			updateMonitoringState(st, monitoredSnaps)
+	defer func() {
+		if err := removeMonitoring(st, snapName); err != nil {
+			logger.Noticef("cannot remove monitoring information: %v", err)
 		}
+	}()
+
+	if aborted {
+		logger.Debugf("monitoring for pre-downloaded snap %q was aborted", snapName)
+		return
 	}
 
-	if continueAutoRefresh {
-		continueInhibitedAutoRefresh(st)
+	// TODO: notify the user that the auto-refresh can't be continued? since this
+	// is a very user-centric flow
+	if err := continueInhibitedAutoRefresh(st, snapName); err != nil {
+		logger.Noticef("cannot continue inhibited auto-refresh for %q: %v", snapName, err)
+		return
 	}
 }
 
-// continueInhibitedAutoRefresh triggers an auto-refresh so it can be continued
-func continueInhibitedAutoRefresh(st *state.State) {
-	// signal that there's an auto-refresh to be continued (for auto-refresh code)
-	st.Cache("auto-refresh-continue-attempt", 1)
-	st.EnsureBefore(0)
-}
+// continueInhibitedAutoRefresh refreshes the snap to continue the inhibited auto-refresh
+func continueInhibitedAutoRefresh(st *state.State, snapName string) error {
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		return fmt.Errorf("cannot get refresh-candidates: %v", err)
+	}
 
-func snapMonitoring(st *state.State) map[string]chan<- bool {
-	if cachedMonitored := st.Cached("monitored-snaps"); cachedMonitored != nil {
-		if monitoredSnaps, ok := cachedMonitored.(map[string]chan<- bool); ok {
-			return monitoredSnaps
-		}
+	hint, ok := refreshHints[snapName]
+	if !ok {
+		return fmt.Errorf("cannot get refresh-candidates for %q: not found", snapName)
+	}
+
+	flags := &Flags{IsAutoRefresh: true, IsContinuedRefresh: true}
+	hint.Flags.IsAutoRefresh = true
+	hint.Flags.IsContinuedRefresh = true
```

Comment:
do we need to do this? aren't the value copied across from the global flags? if they aren't maybe we should look into adjusting that